src/hotspot/share/runtime/continuationFreezeThaw.cpp

```diff
@@ -2018,11 +2191,78 @@ inline bool ThawBase::seen_by_gc() {
   return _barriers || _cont.tail()->is_gc_mode();
 }
 
-NOINLINE intptr_t* ThawBase::thaw_slow(stackChunkOop chunk, bool return_barrier) {
+static inline void relativize_chunk_concurrently(stackChunkOop chunk) {
+#if INCLUDE_ZGC || INCLUDE_SHENANDOAHGC
+  if (UseZGC || UseShenandoahGC) {
+    chunk->relativize_derived_pointers_concurrently();
+  }
+#endif
+}
+
+template <typename ConfigT>
+NOINLINE intptr_t* Thaw<ConfigT>::thaw_slow(stackChunkOop chunk, Continuation::thaw_kind kind) {
+  Continuation::preempt_kind preempt_kind;
+  bool retry_fast_path = false;
+
+  _preempted_case = chunk->preempted();
+  if (_preempted_case) {
+    if (chunk->object_waiter() != nullptr) {
+      // Mounted again after preemption. Resume the pending monitor operation,
+      // which will be either a monitorenter or Object.wait() call.
+      assert(chunk->current_pending_monitor() != nullptr || chunk->current_waiting_monitor() != nullptr, "");
+      ObjectWaiter* waiter = chunk->object_waiter();
+      ObjectMonitor* mon = waiter->monitor();
+      preempt_kind = waiter->is_wait() ? Continuation::freeze_on_wait : Continuation::freeze_on_monitorenter;
+
+      bool mon_acquired = mon->resume_operation(_thread, waiter, _cont);
+      assert(!mon_acquired || mon->is_owner(_thread), "invariant");
+      if (!mon_acquired) {
+        // Failed to aquire monitor. Return to enterSpecial to unmount again.
+        return push_cleanup_continuation();
+      }
+      chunk = _cont.tail();  // reload oop in case of safepoint in resume_operation (if posting JVMTI events).
+    } else {
+      // Preemption cancelled in moniterenter case. We actually acquired
+      // the monitor after freezing all frames so nothing to do.
+      preempt_kind = Continuation::freeze_on_monitorenter;
+    }
+    // Call this first to avoid racing with GC threads later when modifying the chunk flags.
+    relativize_chunk_concurrently(chunk);
+    chunk->set_preempted(false);
+    retry_fast_path = true;
+  } else {
+    relativize_chunk_concurrently(chunk);
```

Comment:
Is the `relativize_chunk_concurrently` solution to the race only to have a single flag read in `can_thaw_fast` or is there some other subtlety here? 

While not required for the PR, if it is just to optimise the `can_thaw_fast` check, it can probably be made to work with one load and still allow concurrent gcs do fast_thaw when we only get here due to a lockstack.

Reply:
Yes, it's just to do a single read. I guess you are thinking of combining flags and lockStackSize into a int16_t?

Reply:
Something along those lines, yes.